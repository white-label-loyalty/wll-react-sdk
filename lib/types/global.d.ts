// Global type definitions for React Native environment

declare const __DEV__: boolean;
