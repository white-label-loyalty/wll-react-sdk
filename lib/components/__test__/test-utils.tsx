import { render } from '@testing-library/react';
import React from 'react';
import { CoreApiUrl, WllSdkProvider } from '../../context/WllSdkContext';
import { defaultTheme } from '../../utils/styling';

const defaultConfig = {
  apiKey: 'test-api-key',
  userToken: 'test-user-token',
  coreApiUrl: CoreApiUrl.STAGING_EU,
};

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <WllSdkProvider
      config={defaultConfig}
      theme={defaultTheme}
      navigationConfig={{}}
    >
      {children}
    </WllSdkProvider>
  );
};

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
