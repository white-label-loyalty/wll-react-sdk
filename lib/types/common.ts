import { ImageProps, ViewStyle } from 'react-native';

export type ImagePropsNoSource = Omit<ImageProps, 'source'>;

export type FlexDirection = 'row' | 'row-reverse' | 'column' | 'column-reverse';

export type Justify =
  | 'start'
  | 'end'
  | 'center'
  | 'between'
  | 'around'
  | 'evenly';

export type Align = 'start' | 'end' | 'center' | 'stretch' | 'baseline';

export type LayoutProps = {
  children: React.ReactNode;
  justify?: Justify;
  align?: Align;
  direction?: FlexDirection;
  style?: ViewStyle;
};

export const justifyMap: Record<Justify, ViewStyle['justifyContent']> = {
  start: 'flex-start',
  end: 'flex-end',
  center: 'center',
  between: 'space-between',
  around: 'space-around',
  evenly: 'space-evenly',
} as const;

export const alignMap: Record<Align, ViewStyle['alignItems']> = {
  start: 'flex-start',
  end: 'flex-end',
  center: 'center',
  stretch: 'stretch',
  baseline: 'baseline',
} as const;
