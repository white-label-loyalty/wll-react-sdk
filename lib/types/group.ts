import { TSection } from './section';

export type TGroup = {
  name: string;
  active: boolean;
  id: string;
  createdAt: string;
  updatedAt: string;
  sections: TSection[];
};
