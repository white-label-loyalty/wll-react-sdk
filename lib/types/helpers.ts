/**
 * Makes a configuration type with required and optional fields,
 * providing default values for optional fields.
 */
export type SafeConfig<T, R extends keyof T> = Required<Pick<T, R>> &
  Partial<Omit<T, R>>;

/**
 * Adds optional children to any component props.
 */
export type WithChildren<T = {}> = T & { children?: React.ReactNode };
