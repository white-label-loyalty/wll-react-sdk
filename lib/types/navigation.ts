export type NavigationType = 'external' | 'internal' | 'modal';

type NavigationHandlerParams = {
  target: string;
  windowTarget: string;
};

export type NavigationConfig = {
  navigationHandlers?: {
    external?: (params: NavigationHandlerParams) => Promise<void>;
    internal?: (params: NavigationHandlerParams) => Promise<void>;
    modal?: (params: NavigationHandlerParams) => Promise<void>;
  };
  external?: (params: NavigationHandlerParams) => Promise<void>;
  internal?: (params: NavigationHandlerParams) => Promise<void>;
  modal?: (params: NavigationHandlerParams) => Promise<void>;
  baseUrl?: string;
};
