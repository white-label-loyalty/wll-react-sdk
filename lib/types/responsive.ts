export type DimensionMode = 'desktop' | 'tablet' | 'mobile';

export type WindowDimensions = {
  width: number;
  height: number;
};

export type ResponsiveContextType = {
  dimensionMode: DimensionMode;
  isTablet: boolean;
  isMobile: boolean;
  isDesktop: boolean;
};
