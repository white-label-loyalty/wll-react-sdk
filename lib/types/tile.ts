export enum CtaAction {
  sameWindow = 'SAME_WINDOW',
  newWindow = 'NEW_WINDOW',
}

export enum BadgeTileType {
  latestEarned = 'LATEST_EARNED',
  specificBadge = 'SPECIFIC_BADGE',
}

export enum TierTileType {
  currentTier = 'CURRENT_TIER',
  nextTier = 'NEXT_TIER',
  specificTier = 'SPECIFIC_TIER',
}

export enum TileType {
  Banner = 'BANNER',
  Points = 'POINTS',
  Content = 'CONTENT',
  Reward = 'REWARD',
  Badge = 'BADGE',
  RewardCategory = 'REWARD_CATEGORY',
  Tier = 'TIER',
}

export enum TileHeight {
  Half = 'HALF',
  Full = 'FULL',
}
type Badge = {
  id: string | null;
  name: string | null;
  description: string | null;
  artworkUrl: string | null;
  createdAt: string | null;
  updatedAt: string | null;
};
export class RewardCategory {
  name?: string | null;
  priority?: number;
  type?: string | null;
  id?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
  description?: string | null;
  metadata?: string | null;
  pictureUrl?: string | null;
  rewards?: Reward[];
  parent?: string | null;
}

export class Reward {
  id?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
  name?: string | null;
  pictureUrl?: string | null;
  description?: string | null;
  value?: number | null;
  price?: number;
  priority?: number;
  discounts?: any[] | null;
  redemptionChannels?: string[] | null;
  venues?: any[] | null;
  purchasableForAudiences?: string[] | null;
  category?: string | null;
  terms?: string | null;
  availability?: Availability;
  purchasable?: boolean;
  tier?: string | null;
  summary?: string | null;
  redemptionMessage?: string | null;
  visibilityCriteria?: string | null;
  logoUrl?: string | null;
  type?: string | null;
  codeType?: string | null;
  code?: string | null;
  purchaseExpiration?: string | null;
  hideCode?: boolean;
  notificationConfig?: string | null;
}
type Availability = {
  start: string | null;
  end: string | null;
};
type Effectivity = {
  start: string | null;
  end: string | null;
};

export type TierType = {
  id: string | null;
  name: string | null;
  description: string | null;
  artworkUrl: string | null;
  priority: number;
  pointsRequirement: number;
  calculation: string | null;
  accumulationPeriod: string | null;
  effectivity: Effectivity;
  createdAt: string | null;
  updatedAt: string | null;
  earnedPoints: number;
  attained: boolean;
};

export class BannerTileConfig {
  imageUrl?: string | null;
  title?: string | null;
  description?: string | null;
  ctaText?: string | null;
  ctaLink?: string | null;
  ctaAction?: CtaAction;
}

export class PointsTileConfig {
  title?: string | null;
  multiplier?: number;
  prefix?: string | null;
  suffix?: string | null;
  imageUrl?: string | null;
  points?: number;
}

export class ContentTileConfig {
  title?: string | null;
  subtitle?: string | null;
  imageUrl?: string | null;
}

export class RewardTileConfig {
  reward?: Reward;
  showPrice?: boolean;
}

export class BadgeTileConfig {
  badgeTileType?: BadgeTileType;
  badgeId?: string | null;
  badge?: Badge;
}

export class RewardCategoryTileConfig {
  categoryId?: string | null;
  allowDecorationOverlay?: boolean;
  rewardCategory?: RewardCategory;
}

export class TierTileConfig {
  tierId?: string | null;
  tier?: TierType;
  type?: TierTileType;
}

export type Tile = {
  id: string | null;
  type: TileType;
  active: boolean;
  createdAt: string | null;
  updatedAt: string | null;
  visibilityCriteria: object | null;
  tileHeight: TileHeight;
  configuration: TileConfig;
};

const getConfigForTileType = (tileType: TileType) => {
  switch (tileType) {
    case TileType.Banner:
      return BannerTileConfig;
    case TileType.Points:
      return PointsTileConfig;
    case TileType.Content:
      return ContentTileConfig;
    case TileType.Reward:
      return RewardTileConfig;
    case TileType.Badge:
      return BadgeTileConfig;
    case TileType.RewardCategory:
      return RewardCategoryTileConfig;
    case TileType.Tier:
      return TierTileConfig;
    default:
      throw new Error(`Tile type ${tileType} is not supported`);
  }
};

export type TileConfig = InstanceType<
  ReturnType<typeof getConfigForTileType>
>;
