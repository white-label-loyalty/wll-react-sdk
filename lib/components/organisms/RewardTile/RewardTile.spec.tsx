import { screen } from '@testing-library/react';
import React from 'react';
import { Tile, TileHeight, TileType } from '../../../types/tile';
import { render } from '../../__test__/test-utils';
import { RewardTile } from './index';
const RewardTileMock: Tile = {
  tileHeight: TileHeight.Full,
  active: true,
  type: TileType.Reward,
  configuration: {
    rewardId: 'f7ce508f-ca52-46ff-bfb7-03e3761feb4a',
    showArtwork: true,
    showDetails: true,
    showPrice: true,
    name: 'Sweet Chilli Lobster Noodles',
    summary: 'Fresh Lobster Noodles with a spicy chilli sauce',
    id: 'f7ce508f-ca52-46ff-bfb7-03e3761feb4a',
    createdAt: '2023-05-05T15:11:42.342Z',
    updatedAt: '2024-07-17T10:45:04.350Z',
    pictureUrl: 'https://ucarecdn.com/a486d015-b4ab-4831-bfa5-68f5bbbf63b1/',
    value: 800,
    price: 10,
    pointsPrefix: '',
    pointsSuffix: '',
    priority: 10,
    availability: {
      start: '2023-05-05T15:10:31.274Z',
      end: '',
    },
    purchasable: true,
    tier: null,
    category: {
      name: 'Chicken Salad',
      priority: 0,
      type: 'REWARD',
      id: '677b6b88-d97c-4b8d-80f4-88b04c86948f',
      createdAt: '2023-05-05T10:39:57.595Z',
      updatedAt: '2023-05-05T10:39:57.595Z',
      description: null,
      metadata: null,
      pictureUrl: 'https://ucarecdn.com/9cda5d3c-75f4-48f0-9f3f-1ece144f6136/',
    },
    discounts: [],
    redemptionMessage: null,
    visibilityCriteria: null,
    type: 'VOUCHER',
    codeType: 'HUMAN',
    code: null,
    purchaseExpiration: null,
    hideCode: false,
    notificationConfig: null,
    artworkUrl: 'https://ucarecdn.com/a486d015-b4ab-4831-bfa5-68f5bbbf63b1/',
    pointsMultiplier: '1',
  },
  id: '39ac25f0-35b3-4668-acba-032dccc63e91',
  createdAt: '2024-10-17T10:50:12.673Z',
  updatedAt: '2024-10-17T10:50:12.673Z',
  priority: 3,
};

describe('<RewardTile />', () => {
  it('renders and matches snapshot', () => {
    const { container } = render(<RewardTile tile={RewardTileMock} />);
    expect(container).toMatchSnapshot();
  });
});

describe('<RewardTile /> Rendering States', () => {
  describe('Size-based rendering', () => {
    it('renders correctly in full size mode', () => {
      const { getByTestId } = render(<RewardTile tile={RewardTileMock} />);

      expect(getByTestId('reward-tile-media')).toBeInTheDocument();
      expect(getByTestId('reward-tile-title')).toBeInTheDocument();
      expect(getByTestId('reward-tile-summary')).toBeInTheDocument();
    });

    it('should not render a half-height tile', () => {
      const halfHeightTile = {
        ...RewardTileMock,
        tileHeight: TileHeight.Half,
      };
      const { container } = render(<RewardTile tile={halfHeightTile} />);
      expect(container).toBeEmptyDOMElement();
    });
  });
});

describe('<RewardTile /> Media', () => {
  it('renders correctly when artworkUrl and showArtwork are provided', () => {
    const { getByTestId } = render(<RewardTile tile={RewardTileMock} />);
    expect(getByTestId('reward-tile-media')).toBeInTheDocument();
  });

  it('does not render when showArtwork is false', () => {
    const tileWithoutShowArtwork = {
      ...RewardTileMock,
      configuration: {
        ...RewardTileMock.configuration,
        showArtwork: false,
      },
    };
    const { queryByTestId } = render(
      <RewardTile tile={tileWithoutShowArtwork} />
    );
    expect(queryByTestId('reward-tile-media')).not.toBeInTheDocument();
  });

  it('does not render media when artworkUrl is missing', () => {
    const tileWithoutArtwork = {
      ...RewardTileMock,
      configuration: {
        ...RewardTileMock.configuration,
        artworkUrl: undefined,
      },
    };
    const { queryByTestId } = render(<RewardTile tile={tileWithoutArtwork} />);
    expect(queryByTestId('reward-tile-media')).not.toBeInTheDocument();
  });
});

describe('<RewardTile /> Points', () => {
  it('renders points when showPrice is true', () => {
    const { getByTestId } = render(<RewardTile tile={RewardTileMock} />);
    expect(getByTestId('reward-tile-points')).toBeInTheDocument();
  });

  it('does not render points when showPrice is false', () => {
    const tileWithoutPrice = {
      ...RewardTileMock,
      configuration: {
        ...RewardTileMock.configuration,
        showPrice: false,
      },
    };
    const { queryByTestId } = render(<RewardTile tile={tileWithoutPrice} />);
    expect(queryByTestId('reward-tile-points')).not.toBeInTheDocument();
  });

  it('calculates points correctly with default multiplier', () => {
    const { getByTestId } = render(<RewardTile tile={RewardTileMock} />);
    const pointsElement = getByTestId('reward-tile-points');
    expect(pointsElement).toHaveAccessibleName('Reward points: 10');
  });

  it('calculates points correctly with custom multiplier', () => {
    const tileWithMultiplier = {
      ...RewardTileMock,
      configuration: {
        ...RewardTileMock.configuration,
        pointsMultiplier: 200,
        pointsPrefix: '',
        pointsSuffix: 'pts',
      },
    };
    const { getByTestId } = render(<RewardTile tile={tileWithMultiplier} />);
    const pointsElement = getByTestId('reward-tile-points');
    expect(pointsElement).toHaveAccessibleName('Reward points: 2000 pts');
  });

  it('uses custom prefix and suffix when provided', () => {
    const tileWithCustomLabels = {
      ...RewardTileMock,
      configuration: {
        ...RewardTileMock.configuration,
        price: 10,
        pointsPrefix: '$',
        pointsSuffix: 'Dollars',
      },
    };
    const { getByTestId } = render(<RewardTile tile={tileWithCustomLabels} />);
    const pointsElement = getByTestId('reward-tile-points');
    const pointsValue = screen.getByTestId('reward-tile-points-value');
    expect(pointsValue.textContent).toBe('10');
    expect(pointsElement).toHaveAccessibleName('Reward points: $ 10 Dollars');
  });

  it('handles empty prefix and suffix correctly', () => {
    const { getByTestId } = render(<RewardTile tile={RewardTileMock} />);
    const pointsElement = getByTestId('reward-tile-points');
    expect(pointsElement).toHaveAccessibleName('Reward points: 10');
  });

  it('handles inactive tile gracefully', () => {
    const mockInactiveTile = {
      ...RewardTileMock,
      active: false,
    };
    const { container } = render(<RewardTile tile={mockInactiveTile} />);
    expect(container.firstChild).toBeNull();
  });
});
