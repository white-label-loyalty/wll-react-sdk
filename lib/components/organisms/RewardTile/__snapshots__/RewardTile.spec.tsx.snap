// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RewardTile /> renders and matches snapshot 1`] = `
<div>
  <button
    aria-label="undefined"
    class="css-view-175oi2r r-cursor-1loqt21 r-touchAction-1otgn73 r-aspectRatio-1ujkv8a r-borderRadius-y47klf r-height-1pi2tsx r-overflow-1udh08x r-position-bnwqim r-width-13qz1uu"
    role="button"
    style="background-color: rgb(248, 247, 252); opacity: 1; flex-direction: column; justify-content: flex-start; align-items: stretch;"
    tabindex="0"
    type="button"
  >
    <div
      class="css-view-175oi2r r-marginBottom-6gpygo r-width-13qz1uu"
      data-testid="reward-tile-media"
      style="flex-basis: 50%;"
    >
      <div
        class="css-view-175oi2r r-alignItems-1awozwy r-justifyContent-1777fci r-bottom-1p0dtai r-height-1pi2tsx r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-width-13qz1uu"
        style="background-color: rgba(57, 46, 215, 0.2);"
      >
        <div
          class="css-view-175oi2r r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af"
          style="background-color: rgba(57, 46, 215, 0.2); opacity: 1;"
        />
        <div
          aria-label="Reward image for Sweet Chilli Lobster Noodles"
          class="css-view-175oi2r r-flexBasis-1mlwlqe r-overflow-1udh08x r-zIndex-417010 r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af"
          style="opacity: 0;"
        >
          <div
            class="css-view-175oi2r r-backgroundColor-1niwhzg r-backgroundPosition-vvn4in r-backgroundRepeat-u6sd8q r-bottom-1p0dtai r-height-1pi2tsx r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-width-13qz1uu r-zIndex-1wyyakw r-backgroundSize-4gszlv"
            style="background-image: url(https://ucarecdn.com/a486d015-b4ab-4831-bfa5-68f5bbbf63b1/);"
          />
          <img
            alt="Reward image for Sweet Chilli Lobster Noodles"
            class="css-accessibilityImage-9pa8cd"
            draggable="false"
            src="https://ucarecdn.com/a486d015-b4ab-4831-bfa5-68f5bbbf63b1/"
          />
        </div>
      </div>
    </div>
    <div
      class="css-view-175oi2r r-alignItems-1habvwh r-flex-13awgt0 r-flexDirection-eqz5dr r-justifyContent-1h0z5md r-paddingBottom-kzbkwu r-paddingInline-3o4zer"
    >
      <div
        class="css-view-175oi2r r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1wtj0ep r-marginBottom-5oul0u r-width-13qz1uu"
      >
        <div
          aria-label="Reward title: Sweet Chilli Lobster Noodles"
          class="css-text-146c3p1 r-maxWidth-dnmrzs r-overflow-1udh08x r-textOverflow-1udbk01 r-whiteSpace-3s2u2q r-wordWrap-1iln25a"
          data-testid="reward-tile-title"
          dir="auto"
          style="color: rgb(0, 0, 0); font-size: 24px; font-weight: bold;"
        >
          Sweet Chilli Lobster Noodles
        </div>
        <div
          class="css-view-175oi2r"
        >
          <svg
            class="lucide lucide-chevron-right"
            fill="none"
            height="24"
            stroke="hsl(0, 0%, 20%)"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m9 18 6-6-6-6"
            />
          </svg>
        </div>
      </div>
      <div
        aria-label="Fresh Lobster Noodles with a spicy chilli sauce"
        class="css-view-175oi2r"
      >
        <div
          class="css-text-146c3p1"
          data-testid="reward-tile-summary"
          dir="auto"
          style="color: rgb(51, 51, 51); font-size: 14px;"
        >
          Fresh Lobster Noodles with a spicy chilli sauce
        </div>
      </div>
    </div>
  </button>
</div>
`;
